### PR TITLE
Fix parameter map access for physical instrument

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Recent changes to position access via the parameter map incorrectly caused the physical instrument to read its positions from `DetectorInfo`. `DetectorInfo` contains only positions for the neutronic instrument. This is fixed here. The solution should be seen as an temporary one, since once Instrument-2.0 has progressed further, it is planned to deal with the physical/neutronic distinction in a different way. However, "temporary" probably will still be a rather long time, so when reviewing please make sure that this solution is reasonable.

**To test:**

- Code review.
- Check that InstrumentView now correctly shows the *physical* instrument instead of the neutronic one (as it used to be). Cannot be tested with BASIS currently, until its IDF is fixed.

No related issue, but this solves parts of #19386 (but also the BASIS IDF needs to be fixed for that).

No release notes, issue was introduced after last release.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Geometry/src/Instrument/Detector.cpp [hunk 1]
```diff
@@ -169,5 +170,13 @@ void Detector::registerContents(ComponentVisitor &componentVisitor) const {
   componentVisitor.registerDetector(*this);
 }
 
+bool Detector::hasDetectorInfo() const {
+  const IComponent *root = this;
+  while (this->m_parent)
```
Comment (OwenArnold) [inline Framework/Geometry/src/Instrument/Detector.cpp]: This looks suspect. Is this supposed to be `root->m_parent` or something similar?

File: Framework/Geometry/src/Instrument/ParameterMap.cpp [hunk 2]
```diff
@@ -1152,14 +1152,20 @@ ParameterMap::create(const std::string &className,
...
-/// contains a DetectorInfo.
-bool ParameterMap::hasDetectorInfo() const {
+/** Only for use by ExperimentInfo. Returns returns true if this instrument
+ contains a DetectorInfo.
+
+ The `instrument` argument is needed for the special case of having a neutronic
+ *and* a physical instrument. `Instrument` uses the same parameter map for both,
+ but the DetectorInfo is only for the neutronic instrument. */
+bool ParameterMap::hasDetectorInfo(const Instrument *instrument) const {
+  if (instrument != m_instrument)
```
Comment (OwenArnold) [inline Framework/Geometry/src/Instrument/ParameterMap.cpp]: Would be good to add a comment around `m_instrument` in the header regarding the fact that `m_instrument` is taken to be the neutronic one.

**Comment by jmborr** [comment]
@SimonHeybrock  I have a BASIS instrument with both neutronic and real space positions which I'll use for testing.

**Comment by jmborr** [comment]
This is BASIS view from branch **master**  using an IDF containing both neutronic and real space positions. It shows the neutronic positions:
before
This is BASIS view from branch **fix_ParameterMap_access_for_physical_instrument** using the previous IDF. It shows the real space positions:
after

Looks good to me.
:shipit:

**Comment by jmborr** [comment]
@OwenArnold Can you merge the pull-request?

**Pull request merged**